platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/transform/AttributeMatchingArtifactVariantSelector.java

```diff
@@ -145,28 +145,14 @@ private static List<TransformedVariant> tryDisambiguate(
         // Find any other candidate which does not match with the last candidate.
         for (int i = 0; i < matches.size() - 1; i++) {
             TransformedVariant current = matches.get(i);
-            if (candidatesDifferent(matcher, current, last)) {
+            if (!matcher.weaklyMatches(current.getAttributes(), last.getAttributes())) {
```

Comment:
Yeah so it seems like you get whats going on here. I think the goals of the method are important to consider, including the complexity of what a more complicated algorithm would be able to achieve. 

The real goal here is to, given a set of transforms chains (in this case of equal length but that does not matter for the algorithm), choose a single transform that is "good". Technically all of these transforms satisfy the users's request, but we are choosing to be a little lenient here and choose one chain that satisfies it, even if more are available. 

What we don't want to do is choose between two different transforms that produce _different_ results. 

Say we have two transform chains: A produces classes, B produces a jar. If the user requested _any_ java code, either chain would work (their shared attributes are weakly compatible), and we would pick either. 

Now, say we have to other chains: C produces a sources zip, D produces a javadoc zip. If the user requested a _zip_, we can't arbitrarily choose between either, since they are not compatible. Based on their attributes, they do not produce something that can be considered "the same".

The existing algorithm solves that for us. It will return a set of size one when all candidates are "same enough". 

In other cases, when there are mismatches, we might return some extras, but that's not quite so big of a deal. Especially if returning a "more accurate set" -- whatever that is defined as -- is more complicated or slower. 

Now, imagine your original example where the current algorithm returns [E, A, C]. It could have returned any other set of results if we picked another "arbitrary" candidate instead of `last`. 

If we chose `B` as our "arbitrary" first candidate, our final set may be different, but we still solve the constraint of returning a set of size one when all candidates weakly match. 

You mention that instead of returning [E, A, C], a more correct algorithm would return [E, A]. But why not [E, C]? Either of these would "work", just as well as [E, A, C] for our original design goal of the algorithm, but to choose between [E, A] or [E, C] we would need to make _another_ arbitrary decision. 

In the end, yes we can technically return a smaller list here. But to get there we would 
1. Increase complexity
2. Make more arbitrary decisions (even making one here is not ideal)

Personally I think this works just fine. Nobody has really complained that we pick the "last" candidate over the "first". We could allow the user to somehow choose a winner if there is more than one candidate transform chain, but again ATM nobody is complaining and that sounds like a lot of work.

Reply:
Thanks for the explanation.  I see you added new javadoc, which is also clearer.  I almost want to say link to this comment chain from it, but I imagine that could get out of date easily.